packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -326,31 +248,104 @@ export const makeZCFZygote = (
-      const zcfSeat = makeZCFSeat(zoeSeatAdmin, seatData);
+  const makeHandleOfferObj = defineDurableKind(handleOfferHandle, () => ({}), {
+    handleOffer: (_context, invitationHandle, seatData) => {
+      const zcfSeat = makeZCFSeat(seatData);
+      // TODO: provide a details that's a better diagnostic for the
+      // ephemeral offerHandler that did not survive upgrade.
       const offerHandler = takeOfferHandler(invitationHandle);
-      const offerResultP = E(offerHandler)(zcfSeat, seatData.offerArgs).catch(
-        reason => {
-          if (reason === undefined) {
-            const newErr = new Error(
-              `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
-            );
-            throw zcfSeat.fail(newErr);
-          }
-          throw zcfSeat.fail(reason);
-        },
-      );
+      const offerResultP =
+        typeof offerHandler === 'function'
+          ? E(offerHandler)(zcfSeat, seatData.offerArgs)
+          : // @ts-expect-error Type issues?
+            E(offerHandler).handle(zcfSeat, seatData.offerArgs);
+
+      const offerResultPromise = offerResultP.catch(reason => {
+        if (reason === undefined) {
+          const newErr = new Error(
+            `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
+          );
+          throw zcfSeat.fail(newErr);
+        }
+        throw zcfSeat.fail(reason);
+      });
       const exitObj = makeExitObj(seatData.proposal, zcfSeat);
       /** @type {HandleOfferResult} */
-      return harden({ offerResultP, exitObj });
+      return harden({ offerResultPromise, exitObj });
     },
   });
+  const handleOfferObj = makeHandleOfferObj();
+
+  const evaluateContract = () => {
+    let bundle;
+    if (passStyleOf(contractBundleCap) === 'remotable') {
+      const bundleCap = contractBundleCap;
+      // @ts-expect-error vatPowers is not typed correctly: https://github.com/Agoric/agoric-sdk/issues/3239
+      bundle = powers.D(bundleCap).getBundle();
+    } else {
+      bundle = contractBundleCap;
+    }
+    return evalContractBundle(bundle);
+  };
+  // evaluate the contract (either the first version, or an upgrade). start is
+  // from a non-upgradeable contract, setupInstallation for upgradeable ones.
+  const { setupInstallation, start, buildRootObject } =
+    await evaluateContract();
+
+  if (start === undefined && setupInstallation === undefined) {
+    assert(
+      buildRootObject === undefined,
+      X`Did you provide a vat bundle instead of a contract bundle?`,
+    );
+    assert.fail(X`unrecognized contract exports`);
+  }
+
+  const contractBaggage = provideDurableMapStore(zcfBaggage, 'contractBaggage');
+
+  // start a non-upgradeable contract
+  const startClassicContract = privateArgs => {
```

Comment:
"classic" is a relative term.

consider "nonupgradable", "static", "fixed" etc

Reply:
Fortunately, much less of an issue now! I'm not even sure we still need terminology for this.